activesupport/lib/active_support/core_ext/securerandom.rb

```diff
@@ -36,10 +36,7 @@ def self.base58(n = 16)
   #   p SecureRandom.base36 # => "4kugl2pdqmscqtje"
   #   p SecureRandom.base36(24) # => "77tmhrhjfvfdwodq8w7ev2m7"
   def self.base36(n = 16)
-    SecureRandom.random_bytes(n).unpack("C*").map do |byte|
-      idx = byte % 64
-      idx = SecureRandom.random_number(36) if idx >= 36
-      BASE36_ALPHABET[idx]
-    end.join
+    n = n ? n.to_int : 16
+    choose(BASE36_ALPHABET, n)
```

Comment:
I'm not sure we really care about performance here... but if we do, this seems trivially faster:

```suggestion
    SecureRandom.random_bytes(n).unpack("C*").map do |byte|
      byte = SecureRandom.random_number(36) if byte >= 252
      BASE36_ALPHABET[byte % 36]
    end.join
```

(Perhaps that's arguably a potential optimisation in `choose` itself, but still, I think we'd also want to be sure it's officially exposed before we start depending on it.)

Reply:
I mentioned this in the Ruby issue tracker after creating this PR.
There already was an existing request to make it public which has now been merged:
https://github.com/ruby/ruby/pull/8312

Reply:
Also:

```ruby
>> SecureRandom.random_bytes(30).unpack("Q*").map { |b| b.to_s(32) }.join
=> "3l60d983j1g885q1hjs688ebs2e0taglo6effo"
```

I haven't benchmarked but I suspect it blows having to make one lookup per character.

Reply:
```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "activesupport"#, github: "rails/rails", branch: "main"
  gem "benchmark-ips"
end

require "active_support"
require "active_support/core_ext/securerandom"

module SecureRandom
  def self.fast_base36(n)
    n = n ? n.to_int : 16
    choose(BASE36_ALPHABET, n)
  end

  def self.simple_base36(n)
    random_bytes(n).unpack("Q*").map { |b| b.to_s(32) }.join
  end
end

puts RUBY_VERSION
[10, 100, 1000, 10000].each do |length|
  puts
  puts " #{length} ".center(80, "=")
  puts

  Benchmark.ips do |x|
    x.report("base36")      { SecureRandom.base36(length) }
    x.report("fast_base36") { SecureRandom.fast_base36(length) }
    x.report("simple_base36") { SecureRandom.simple_base36(length) }
    x.compare!(order: :baseline)
  end
end
```

```
3.1.2

====================================== 10 ======================================

Warming up --------------------------------------
              base36    18.414k i/100ms
         fast_base36    20.125k i/100ms
       simple_base36    79.170k i/100ms
Calculating -------------------------------------
              base36    194.352k (± 4.7%) i/s -    975.942k in   5.033364s
         fast_base36    200.712k (± 7.9%) i/s -      1.006M in   5.052538s
       simple_base36    793.331k (± 5.1%) i/s -      3.958M in   5.003052s

Comparison:
              base36:   194352.0 i/s
       simple_base36:   793331.1 i/s - 4.08x  faster
         fast_base36:   200711.6 i/s - same-ish: difference falls within error


===================================== 100 ======================================

Warming up --------------------------------------
              base36     2.256k i/100ms
         fast_base36     2.612k i/100ms
       simple_base36    14.970k i/100ms
Calculating -------------------------------------
              base36     22.888k (± 6.7%) i/s -    115.056k in   5.058205s
         fast_base36     26.313k (± 3.4%) i/s -    133.212k in   5.068967s
       simple_base36    146.243k (± 7.6%) i/s -    733.530k in   5.051720s

Comparison:
              base36:    22888.1 i/s
       simple_base36:   146242.9 i/s - 6.39x  faster
         fast_base36:    26312.9 i/s - 1.15x  faster


===================================== 1000 =====================================

Warming up --------------------------------------
              base36   244.000  i/100ms
         fast_base36   268.000  i/100ms
       simple_base36     1.682k i/100ms
Calculating -------------------------------------
              base36      2.433k (± 6.7%) i/s -     12.200k in   5.048177s
         fast_base36      2.743k (± 3.8%) i/s -     13.936k in   5.088650s
       simple_base36     17.775k (± 9.8%) i/s -     90.828k in   5.220362s

Comparison:
              base36:     2432.6 i/s
       simple_base36:    17775.5 i/s - 7.31x  faster
         fast_base36:     2742.7 i/s - 1.13x  faster


==================================== 10000 =====================================

Warming up --------------------------------------
              base36    23.000  i/100ms
         fast_base36    25.000  i/100ms
       simple_base36   185.000  i/100ms
Calculating -------------------------------------
              base36    231.261  (±13.0%) i/s -      1.127k in   5.006034s
         fast_base36    273.471  (± 4.8%) i/s -      1.375k in   5.039957s
       simple_base36      1.739k (± 8.3%) i/s -      8.695k in   5.040199s

Comparison:
              base36:      231.3 i/s
       simple_base36:     1738.9 i/s - 7.52x  faster
         fast_base36:      273.5 i/s - same-ish: difference falls within error
```

Reply:
Unfortunately, it doesn't use the same alphabet:

```
>> 32.times.map {|i| i.to_s(32)}
=> 
["0",                                                 
 "1",                                                 
 "2",                                                 
 "3",                                                 
 "4",                                                 
 "5",                                                 
 "6",                                                 
 "7",                                                 
 "8",                                                 
 "9",                                                 
 "a",                                                 
 "b",                                              
 "c",
 "d",
 "e",
 "f",
 "g",
 "h",
 "i",
 "j",
 "k",
 "l",
 "m",
 "n",
 "o",
 "p",
 "q",
 "r",
 "s",
 "t",
 "u",
 "v"]
```

Not sure if it's a big deal or not.

Reply:
> it doesn't use the same alphabet:

Scratch that, I misread it's the same.

Reply:
This doesn't seem to work for small `n`.
```ruby
SecureRandom.random_bytes(5).unpack("Q*").map { |b| b.to_s(32) }.join
=> ""
```

It would also need to be sliced as the length can be more than 1 character:
```ruby
SecureRandom.random_bytes(100).unpack("Q*").map { |b| b.to_s(32) }.join.size
=> 152
```

Reply:
> This doesn't seem to work for small n.

Right, good point, it only works with multiple of 8bytes of data. I guess:

```
.unpack("Q*S*C*")
```

Works.

> It would also need to be sliced as the length of the unpacked byte can be more than 1 character:

I don't follow.

Reply:
Ah nevermind, yeah, it would need to be padded.